msprime/mutations.py

```diff
@@ -56,30 +55,69 @@
 ]
 
 
-# NOTE we're doing this hack of monkey-patching asdict onto the
-# MutationModel subclasses because of a bug in sphinx, where it's
-# not possible to have muliple inheritance with mocked out classes.
-def safe_asdict(self):
-    # This version of asdict makes sure that we have sufficient parameters
-    # to call the contructor and recreate the class. However, this means
-    # that subclasses *must* have an instance variable of the same name.
-    # This is essential for Provenance round-tripping to work.
-    return {
-        key: getattr(self, key)
-        for key in inspect.signature(self.__init__).parameters.keys()
-        if hasattr(self, key)
+def mutation_model_factory(model):
+    """
+    Returns a mutation model corresponding to the specified model.
+    - If model is None, the default mutation model is returned.
+    - If model is a string, return the corresponding model instance.
+    - If model is an instance of MutationModel, return a copy of it.
+    - Otherwise raise a type error.
+    """
+
+    model_map = {
+        # "slim": SlimMutationModel(), Needs type argument so can't be string init'd
+        "infinite_alleles": InfiniteAllelesMutationModel(),
```

Comment:
This list really illustrates part of the problem here - we have 3 different conventions for naming these classes. Should they all have the suffix "MutationModel"? I guess this isn't so bad when using the short names, so we'd have 
```
InfiniteAllelesMutationModel
SlimMutationModel
BinaryMutationModel
JC69MutationModel
HKYMutationModel
F84MutationModel
BLOSUM62MutationModel
PAMMutationModel
```

I guess we could also use the suffix "Mutations" which would be a little less wordy (but I guess a little less precise too).

Reply:
It's not so bad that the class names are a bit long, as we'd not be using them very often (we'd mostly use the strings).

Reply:
> I don't think we have much compatibility baggage here.

Only `InfiniteSites` existed in the last release, so yeah lets deprecate that and get this exactly how we want it.

Reply:
I've gone with *MutationModel in https://github.com/tskit-dev/msprime/pull/1176/commits/29846b3cc5787d49396ccf6246b5d7e870377e28